### ISSUE TITLE
Stockholm parser ignores per-column attributes e.g. GC

### ISSUE THREAD

**Issue by wwood** [biopython/biopython#357]
Hi,

I'm writing an app that requires the `GC` row of a stockholm format output, but as far as I can tell, biopython simply ignores this part of the file. Perhaps I am misinterpreting or misusing biopython somehow?

From https://github.com/biopython/biopython/blob/c83880bc304a14006f534bd57028e1916fe559ce/Bio/AlignIO/StockholmIO.py#L379

``` python
elif line[:5] == '#=GC ':
    #Generic per-Column annotation, exactly 1 char per column
    #Format: "#=GC <feature> <exactly 1 char per column>"
    pass
```

I'm not familiar enough with the biopython code base to suggest how this might be exposed, other than 'somehow' :)

Another point of information loss is that all `.` characters are converted to `-` in the alignments, when these have distinct meanings.

``` python
seqs[id] += seq.replace(".", "-")
```

Here's a full example file:

```
# STOCKHOLM 1.0

#=GS contig_24650_5 DE # 4239 # 5012 # -1 # ID=357_5;partial=00;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.574
#=GS contig_35394_2 DE # 1679 # 2557 # 1 # ID=436_2;partial=00;start_type=ATG;rbs_motif=GGAGG;rbs_spacer=5-10bp;gc_cont=0.573
#=GS contig_4083_12 DE # 7484 # 8395 # -1 # ID=477_12;partial=00;start_type=ATG;rbs_motif=GGAG/GAGG;rbs_spacer=5-10bp;gc_cont=0.565

contig_24650_5         ..................................mkkkiisian------QKGGVGKTTTAINLAASLAAAEKRTLLIDSDPQGNSSS----------------------------------------------------------------------------------------------------------------
#=GR contig_24650_5 PP ..................................**98877665......59******************************999876................................................................................................................
contig_35394_2         mscnnehgrcnegecgnsgsapqkkadagnenierrmsrirhki--MVLSGKGGVGKSTVAANLAIALALEGNRVGLLDVDFH---------------------------------------------------------------------------------------------------------------------
#=GR contig_35394_2 PP *****************************************966..4678***********************9988777755.....................................................................................................................
contig_4083_12         .mdqasglrmrmekgkagmfrespptavpgkgrehhavrviavt-----SGKGGVGKTNLVANLACLLSRMKKKVLLLDGDTGLANIDVVLGLSPRYNLSHVLREE----CTLSDALIKGPGGFLILPSSSGIQEMSDLTRE--QKLVILEELNHLEMDLDYMLIDTAAGIA--GNVMYF-NLAAREIIVVASPEPTSLT
#=GR contig_4083_12 PP .*************************************876544.....59*****************************************************99....79***********9998877666667667776..456679****************877654..555554.45799**************
#=GC PP_cons           ..............................................46778************************9999998899988****************99....79***********9998877666667667776..456679****************877654..555554.45799**************
#=GC RF                ............................................xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

contig_24650_5         --------------------------.----------------......-----------------------------------------------------------------------------gvgidrtkireknlyqaiigkipvaetlhasplpsltvipatqdligvevefvnlsdreqrlkqllsglpdnfe
#=GR contig_24650_5 PP ..............................................................................................................................567778889999999***********************************************************
contig_35394_2         --------------------------.----------------......-----------------------------------------------------------------------------gpsiptllnlegrqlsydgnsispvrfsdnlsvvslgfllrnrddaviwrgpmkmnvirqilsetewgdldfli
#=GR contig_35394_2 PP ..............................................................................................................................5555666677888999**********************************************************
contig_4083_12         DAYALIKIL--YQRQGKKRFR-LIANmVQNAAEGMEVFNRLNQatehflNLSVEYLGHIVRDESLPRAVRQQKAAVEVYPQAPAVKCMERIAHKIIKEK---------------------------pdcglngsisffwnqdivqggr*...................................................
#=GR contig_4083_12 PP ******977..5566666554.344413445555666665544111222567788999*************************************9875...........................678999*****************...................................................
#=GC PP_cons           ******977..5566666554.3444.3445555666665544......567788999*************************************9875.....................................................................................................
#=GC RF                xxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx......xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx..........................................................................

contig_24650_5         fviidcppslgfltinalvasdsliiplqceyfameglsqllntvrlvqtrlnpglilsgilltmfdtrnrlshrveeevrkyfagkvfetiiprnvrlsespshglpvllyditsrgavsymelageilnngrr*..
#=GR contig_24650_5 PP ****************************************************************************************************************************************..
contig_35394_2         idfppgtgdeplsfaqlipgsdgavivttpqnlsvidvrksinfcrqvhvpvlgvienmsgffcphcgkeahlfkpgggeamakdmgvpflgrvpvdpmiaeasdngkpfiyfsgktetgkafsrivekiearekav*
#=GR contig_35394_2 PP ******************************************************************************************************************************************
contig_4083_12         ..........................................................................................................................................
#=GR contig_4083_12 PP ..........................................................................................................................................
#=GC PP_cons           ..........................................................................................................................................
#=GC RF                ..........................................................................................................................................
//
```

Thanks!
ben

**Comment by peterjc** [comment]
See also https://www.biostars.org/p/137106/#138253 which is the same problem about ignoring the ``pp_cons`` per-column information.

**Comment by peterjc** [comment]
I've got something ready for comment/testing: pull request #528.

This would expose the ``PP_cons`` and ``RF`` type ``GC`` lines as follows:

```
>>> from Bio import AlignIO
>>> align = AlignIO.read("example.sth", "stockholm")
>>> print(align.column_annotations.keys())
dict_keys(['posterior_probability', 'reference_annotation'])
>>> align.column_annotations['posterior_probability']
'..............................................46778************************9999998899988****************99....79***********9998877666667667776..456679****************877654..555554.45799********************977..5566666554.3444.3445555666665544......567788999*************************************9875...............................................................................................................................................................................................................................................'
>>> align.column_annotations['reference_annotation']
'............................................xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx......xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx....................................................................................................................................................................................................................'
```

You would be able to slice, e.g. taking selected columns:

```
>>> new = align[:,20:30] + align[:,35:40]
>>> new.column_annotations['reference_annotation']
'...............'
>>> new.column_annotations['posterior_probability']
'...............'
>>> align.column_annotations['posterior_probability'].find("7")
48
>>> new = align[:,40:50] + align[:,70:75]
>>> new.column_annotations['reference_annotation']
'....xxxxxxxxxxx'
>>> new.column_annotations['posterior_probability']
'......4677*****'
>>> print(new)
SingleLetterAlphabet() alignment with 3 rows and 15 columns
sian------AEKRT contig_24650_5
rhki--MVLSEGNRV contig_35394_2
iavt-----SMKKKV contig_4083_12
```

And you could also output the recognised ``GC`` lines:

```
>>> print(new.format("stockholm"))
# STOCKHOLM 1.0
#=GF SQ 3
contig_24650_5 sian------AEKRT
#=GS contig_24650_5 AC contig_24650_5
#=GS contig_24650_5 DE # 4239 # 5012 # -1 # ID=357_5;partial=00;start_type=ATG;rbs_motif=None;rbs_spacer=None;gc_cont=0.574
#=GR contig_24650_5 PP 7665......*****
contig_35394_2 rhki--MVLSEGNRV
#=GS contig_35394_2 AC contig_35394_2
#=GS contig_35394_2 DE # 1679 # 2557 # 1 # ID=436_2;partial=00;start_type=ATG;rbs_motif=GGAGG;rbs_spacer=5-10bp;gc_cont=0.573
#=GR contig_35394_2 PP *966..4678***99
contig_4083_12 iavt-----SMKKKV
#=GS contig_4083_12 AC contig_4083_12
#=GS contig_4083_12 DE # 7484 # 8395 # -1 # ID=477_12;partial=00;start_type=ATG;rbs_motif=GGAG/GAGG;rbs_spacer=5-10bp;gc_cont=0.565
#=GR contig_4083_12 PP 6544.....5*****
#=GC PP_cons ......4677*****
#=GC RF ....xxxxxxxxxxx
//
```

**Comment by peterjc** [comment]
Should have closed this two years ago with the merging of #528.
